src/LinkedGenome.py

```diff
@@ -0,0 +1,192 @@
+        self.val = val
+        self.prev = p
+        self.next = n
+
+def insert_after(link: Link[T], val: T) -> None:
+    """Add a new link containing avl after link."""
+    new_link = Link(val, link, link.next)
+    new_link.prev.next = new_link
+    new_link.next.prev = new_link
+def insert_before(link: Link[T], val: T) -> None:
+    """Add a new link containing avl after link."""
+    new_link = Link(val, link, link.prev)
+    new_link.prev.next = new_link
+    new_link.next.prev = new_link
+
+class LinkedListGenome(Genome):
+    """
+    Representation of a genome.
+
+    Implements the Genome interface using linked lists.
+    """
+    head: Link[Feature]  # Dummy head link
+    empty_te, active_te, inactive_te = 0, 1, 2
+    active_identifier = {}
+    counter_te = 0
+
+    def __init__(self, n: int):
+        """Create a new genome with length n."""
+        self.head = Link(None, None, None)  # type: ignore
+        self.head.prev = self.head
+        self.head.next = self.head
+        insert_after(self.head.prev, Feature(self.empty_te, n))
+    
+    def __iter__(self):
+        feature = self.head.next
+        while feature is not self.head:
+            yield feature.val
+            feature = feature.next
+    def into_iter_with_pos(self):
+        feature = self.head.next
+        acc = 0
+        while feature is not self.head:
+            acc += feature.val.length 
+            yield (feature, acc)
+            feature = feature.next
+
+    def insert_te(self, pos: int, length: int) -> int:
+        """
+        Insert a new transposable element.
+
+        Insert a new transposable element at position pos and len
+        nucleotide forward.
+
+        If the TE collides with an existing TE, i.e. genome[pos]
+        already contains TEs, then that TE should be disabled and
+        removed from the set of active TEs.
+
+        Returns a new ID for the transposable element.
+        """
+
+        for feature, end in self.into_iter_with_pos():
+            if end > pos:
+                self.insert_into(length, feature, feature.val.length -end + pos, end - pos)
+                return self.counter_te
+        return -1
+
+    def insert_into(self, length, feature, split_size_1, split_size_2):
+        if  feature.val.feature == self.active_te:
+            self.disable_feature(feature)
+        feature.val = Feature(feature.val.feature, split_size_1)
+        insert_after(feature, Feature(self.active_te, length))
+        insert_after(feature.next, Feature(feature.val.feature, split_size_2))
+        self.counter_te += 1
+        self.active_identifier[self.counter_te] = feature.next
+
+    def disable_feature(self, feature):
+        feature.val = Feature(self.inactive_te, feature.val.length)
+        for identifier, disable in self.active_identifier.items():
+            if disable is feature:
+                self.active_identifier.pop(identifier)
+                break
```

Comment:
Why would you do a linear search here? If you have a feature's ID, you can get it as `self.active_identifier[ID]`. You could disable a TE in constant time instead of $O(k)$.

Reply:
Yes, absolutely right